File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/mutation/NestedMutationIntent.java
package com.foundgine.core.semantic.planning.mutation;

import java.util.*;

public record NestedMutationIntent(IMutationIntent mutation, List<NestedMutationChild> children) {
    public NestedMutationIntent {
        Objects.requireNonNull(mutation);
        children = children == null ? List.of() : List.copyOf(children);
    }
}

File: src/csharp/Foundgine.Core/Semantic/Planning/Mutation/NestedMutationIntent.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Planning.Mutation;

/// <summary>
/// A mutation tree. The root mutation is followed by child mutations attached
/// through metadata relationships. The planner flattens the tree into the
/// existing dependency-aware mutation batch; SQL remains unaware of the tree.
/// </summary>
public sealed record NestedMutationIntent(
    IMutationIntent Mutation,
    IReadOnlyList<NestedMutationChild> Children);

public sealed record NestedMutationChild(
    RelationshipId Relationship,
    NestedMutationIntent Mutation);